[1단계 - 로또 구현] 완태(김태완) 미션 제출합니다.

wannte opened the pull request:
안녕하세요! 완태입니다:)
만나서 반갑습니다.

이번 미션 잘 부탁드립니다!!

프로그램 작성 후 학습 회고를 작성했는데요, 제가 했던 부분을 정리한 것이라, 다소 건방진 말투를 사용했습니다. 이 부분 미리 죄송합니다 ㅎ

학습회고 부분에서 제가 고민했던 부분, 질문들이 있어 이와 관련하여 코멘트 해주시면 정말 감사하겠습니다!!

---

# 학습 회고


## Enum을 사용하여 등수 정보를 저장 - 5

1. 등수 안에 들지 못했을 때를 MISS라고 따로 두어 등수를 관리. 

→ 장점 : 등수 안에 들지못했을 때의 예외 처리를 안 해도 됨

→ 단점 : 결과값 출력 부분에서는 MISS의 예외 처리를 해줘야 함.

(개인적으로, 등수에 들지 않는 로또의 갯수도 활용가능성이 있다고 생각하여, 이와 같이 구현)

 2. Enum의 구성에 대한 고민

static getRating함수를 통해 해당하는 등수값을 반환하는데, 각 enum이 로또 일치 번호수, 보너스볼 일치 여부의 정보를 가지고 있게 끔 구현하는 가에 대한 고민.

현재 방식: matchCount, reward만을 저장. 2등인 경우만, containBaseBall 조건을 추가하여 hardCoding.

다른 방식 : 로또의 겹치는 값과 보너스볼 여부를 Result로 묶어 저장하고 이를 List로 관리.

```java
Enum Rating{
    private List<Result> result;
    private int reward)
}

class Result {
    private int matchCount;
    private boolean containBaseBall;
}
```

4등에 경우 (4,false), (f,true)의 경우를 모두 포함해야 해서, Result 결과를 List로 관리해야겠다는 생각을 했음.

→ 이렇게 했을 때, 현재 getRating의 하드코딩 되어 있는부분을 줄일 수 있음.

(이 방식으로 한번 시도를 해봐야겠음! TODO) ->  'refactor: Rating 구조 개선' 이후에는 이를 수정한 상태입니다.

 _페어와의 논의 결과, 너무 많은 복잡성을 Enum에 부여하는 것 같아, matchCount와 reward만으로 구성된 Rating을 사용_


[https://github.com/wannte/java-lotto/blob/step1/src/main/java/lotto/domain/rating/Rating.java](https://github.com/wannte/java-lotto/blob/step1/src/main/java/lotto/domain/rating/Rating.java)

## 생성자의 dummy 사용 - 3

List<Integer> List<LottoNumber>를 동시에 사용했을 때, Erase~~ 에러를 띄움. 기존에는 static createByName(List<Integer> numbers)와 같은 방식으로 구현을 진행했는데, JDK HashSet 의 생성자에 경우 dummy 를 추가하는 방식으로 생성자 오버로딩을 구현해서 이런식으로 구현을 진행.

현업에서는 어떻게 사용되는지 궁금합니다!

참고링크

[https://github.com/wannte/java-lotto/blob/step1/src/main/java/lotto/domain/lotto/Lotto.java](https://github.com/wannte/java-lotto/blob/step1/src/main/java/lotto/domain/lotto/Lotto.java)

## 함수형 인터페이스 with 전략 패턴 (LottoMachine) - 3

저번 랜덤값의 테스트를 진행할 때와 마찬가지로, 이번과제에서도 shuffle를 사용하여, 1~45 중 랜덤하게 로또 번호를 생성하기 때문에, 이를 테스트하기 위해서 랜덤하지 않을 값을 통해 이를 진행해야했다.

LottoMachine이라는 인터페이스를 만들고 이를 상속받는 RandomLottoMachine 그리고 DefaultLottoMachine(항상 1,2,3,4,5,6의 로또 생성) 을 구현하고 진행하려고 하였으나, 함수형 인터페이스를 이용해 이를 따로 구현하지 않고 테스트를 진행할 수 있었다.

참고링크

 [구현 코드](https://github.com/wannte/java-lotto/tree/step1/src/main/java/lotto/domain/lottomachine)
 [테스트](https://github.com/wannte/java-lotto/blob/step1/src/test/java/lotto/domain/lotto/LottoRepositoryTest.java)

## Domain 배치에 관한 고민 - 2

처음 구현을 진행했을 때, domain 내부적으로는 패키지 구분없이 클래스를 계속 생성하여 프로젝트를 진행.

이를 분리를 시킬때, 그 기준을 정하는데에 어려움을 겪음.

현재 구성상, 기능적으로 묶이는 것들로 묶고, 특정기능에 속하기 애매한 Primitive타입을 구분.

이와 관련한 가이드라인이나  레퍼런스를 찾아 공부를 진행해야겠음!

[https://github.com/wannte/java-lotto/tree/step1/src/main/java/lotto/domain](https://github.com/wannte/java-lotto/tree/step1/src/main/java/lotto/domain)

pjy1368 committed:
docs: 요구 사항 정의 및 커밋 컨벤션 작성

pjy1368 committed:
refactor: 사용하지 않는 파일 삭제 및 init

pjy1368 committed:
docs: 로또 유효성 검사 기능 추가

pjy1368 committed:
feat(lotto): 로또 유효성 검사 기능 구현

pjy1368 committed:
feat(LottoSupplier): 로또 번호 생성 기능 구현

pjy1368 committed:
docs: 구현한 기능 체크

pjy1368 committed:
refactor(Lotto): 접근제한자 수정

pjy1368 committed:
refactor(LottoSupplier): 상수 처리

pjy1368 committed:
docs: 구현한 기능 체크

pjy1368 committed:
feat: 구매금액에 따라 티켓 수를 구하는 기능 구현

pjy1368 committed:
docs: 구현한 기능 체크

pjy1368 committed:
feat(Lotto): 일치하는 로또 번호 개수를 구하는 기능 구현

pjy1368 committed:
feat(Lotto): 보너스볼 포함 여부 확인하는 기능 구현

pjy1368 committed:
refactor(LottoNumber): 원시값을 포장함

pjy1368 committed:
docs: to-do List 업데이트

pjy1368 committed:
sytle(LottoTest): 오타 수정

pjy1368 committed:
feat(Rating): 등수 구하는 기능 구현

pjy1368 committed:
feat(RatingInfo): 등수를 세는 기능 구현

pjy1368 committed:
feat(Money, Ticket): 구입금액에 따른 복권 개수 구하는 기능 구현

pjy1368 committed:
style(RatingInfoTest): 오타 수정

pjy1368 committed:
feat: 로또 긁고 등수를 세는 기능 구현

pjy1368 committed:
feat(LottoStats): 당첨 내역을 문자열로 반환하는 기능 구현

pjy1368 committed:
feat(LottoStats): 수익률 구하는 기능 구현

pjy1368 committed:
refactor: 패키지 구조 변경

pjy1368 committed:
feat(InputView): 입력 로직 구현

pjy1368 committed:
feat(OutputView): 출력 로직 구현

pjy1368 committed:
feat(WinningLotto): 보너스 볼 유효성 검사 기능 구현

pjy1368 committed:
feat(Application): 컨트롤러 로직 구현

pjy1368 committed:
fix(Money): 구매금액의 입력이 0일 때 예외 처리

pjy1368 committed:
refactor(LottoMachine): LottoSupplier에서 클래스명 변경

pjy1368 committed:
refactor(InputView): 입력 공백 예외 처리

pjy1368 committed:
refactor(LottoRepository): 로또 객체를 담는 리스트를 일급콜렉션 사용하여 변경

pjy1368 committed:
refactor: 패키지 구조 수정

pjy1368 committed:
refactor(view): 출력 메시지 수정

pjy1368 committed:
refactor(LottoController): 컨트롤러 분리

pjy1368 committed:
refactor: 하드코딩된 부분 상수 처리

pjy1368 committed:
style: 접근제한자 추가

pjy1368 committed:
docs: 불필요한 텍스트 제거

pjy1368 committed:
fix(Ticket): 돈이 1000원 단위가 아니더라도 복권 구매 가능하도록 변경

pjy1368 committed:
style: 에러 메시지 수정

pjy1368 committed:
refactor(Counter): 당첨된 등수의 횟수를 불변객체로 생성

pjy1368 committed:
test(LottoManagerTest): butLotto 테스트 코드 작성

pjy1368 committed:
test(LottoRepositoryTest): 로또 생성 정보 비교 테스트 코드 작성

pjy1368 committed:
test(LottoManagerTest): 로또 긁는 테스트 코드 작성

pjy1368 committed:
style: DisplayName 추가

pjy1368 committed:
fix(LottoStatistics): 수익률을 double로 변경

wannte committed:
style: import, line optimize

wannte committed:
refactor: ratingInfo -> ratingCounter

wannte committed:
refactor: Rating 구조 개선

LottoResult 를 추가하여, 결과값 포장.
생성과정에서 부터 isSecond 요소를 추가하여 Rating 을 간단하게
TODO: isSecond 메소드 명을 포괄적으로 수정. 5 hard 코딩 수정

Hue9010 commented on src/main/java/lotto/domain/statistics/LottoStatistics.java:
```diff
@@ -0,0 +1,56 @@
+    private static final String ENTER = System.lineSeparator();
+    private static final String START_PRINT_FORMAT = "%d개 일치";
+    private static final String HAS_BONUS_BALL = ", 보너스 볼 일치";
+    private static final String END_PRINT_FORMAT = " (%d원) - %d개" + ENTER;
```
많이들 하는 실수군요. 😉 
```java
        stringBuilder.append(String.format(END_PRINT_FORMAT,rating.getReward(),count));
```
stringBuilder에서 사용되고 있습니다. 그런데 + 연산이 섞여 있기 때문에 stringBuilder를 사용한 목적이 퇴색됐습니다. 😢 

사실 더하기 연산도 컴파일 단에서 최적화를 해준답니다.
[dk1.5에서 String 더하기의 컴파일시의 최적화](https://gist.github.com/benelog/b81b4434fb8f2220cd0e900be1634753)

Hue9010 commented on src/main/java/lotto/domain/statistics/LottoStatistics.java:
```diff
@@ -0,0 +1,56 @@
+    private static final String ENTER = System.lineSeparator();
+    private static final String START_PRINT_FORMAT = "%d개 일치";
+    private static final String HAS_BONUS_BALL = ", 보너스 볼 일치";
+    private static final String END_PRINT_FORMAT = " (%d원) - %d개" + ENTER;
```
다른 곳도 포함인데 이렇게 모든걸 상수로 빼는게 유용하던가요? 🤔 
리뷰어마다 의견이 다를 부분일텐데 재활용성, 매직넘버로서의 의미를 가지지 않는다면 불필요한 상수라 판단되어 별도로 상수화 시키지 않습니다.

```java
            stringBuilder.append(HAS_BONUS_BALL);
            stringBuilder.append(", 보너스 볼 일치");
```
모드상에 `"`가 많이 나타나면 마음에 안드는 부분이 있긴하지만 🙄 상수화를 통해 도리어 어떤 값이 사용되고 있는지 한눈에 파악하기 어려워 지지 않았나 생각듭니다.
이건 앞서 말씀드린 것 처럼 리뷰어마다 스타일이 다를 수 있어 완태는 어떠셨는지 의견을 듣고 싶어요. 🙏

그리고 ENTER보다는 NEW_LINE같이 개행의 의미를 넣어주는게 더 좋아 보입니다. 😄

wannte replied:
안녕하세요 휴!
사실 이 부분에 대해서는 제 기준이 확실하게 잡혀있지 않아 궁금했던 부분입니다!
지난 번, 자동차 구현 때에는 휴의 말처럼 반복되는 부분이 아니라면, 상수화를 진행하지 않았었습니다!

이번 페어는 이 부분에 대해서는 대부분 상수화를 진행하는 편이였는데, 딱히 저의 기준이 없어서 그대로 진행하였습니다.

지금와서 제 생각을 정리해 보면, 모드 상에 "" 를 사용하는 것이, 별개로 상수를 빼서 따로 관리하는 것보다는 좋다는 생각이드네요!!

NEW_LINE의 네이밍이 훨씬 좋은 것 같네요 감사합니다:)

Hue9010 replied:
아직 기준이 안잡혀있다면 우선 모두 상수화해보는 것도 방법입니다.
고기도 먹어 본 사람이 맛을 안다고... 😋
비슷하게 저도 스트림에 익숙치 않은 분이 스트림 학습을 하시고자 한다면 반복문을 사용하는게 나아보이는 곳도 스트림을 사용해보라고 조언을 드리기도 합니다.
생각해보니 상수화를 과도하게 안하는 이유 중 하나가 변수명 짓는 시간이 오래 걸리는 것도 한 몫 하네요. 🤣

wannte replied:
네 사실 저는 그 상수화 과정이 너무 싫습니다.. ㅎㅎ
지금 와서는 intellij 커맨드를 알아서 편하게 상수로 만들수있지만요..

Hue9010 commented on src/main/java/lotto/domain/statistics/LottoStatistics.java:
```diff
@@ -0,0 +1,56 @@
+        }
+        stringBuilder.append(String.format(END_PRINT_FORMAT,rating.getReward(),count));
+        return stringBuilder.toString();
+    }
```
결국 OutputView에서의 출력 문구를 위한 작업이네요.
어떤식으로 보여줄 지에 대한 표현은 View에다가 위임하는걸 추천드려요.
View에는 데이터만 전달해주고 그 데이터를 어떻게 표현할지는 View에서 자체적으로 판단해서 보여주게 말이죠.

웹페이지를 예시로 들면 animal이라는 필드에 pig라는 데이터가 전달되면 이 데이터를 어떻게 보여줘야 할지에 대한 것은 View에서 결정짓고 처리하면 됩니다. 표현이라해서 어려운게 아니라 pig라는 데이터를 돼지 그림으로 보여줄지 돼지, pig, Pig, PIG 등등 어떤 형태로 보여줄지의 문제이죠.
그러한 표현에 대한 처리가 여기서 이뤄지고 있는걸로 보입니다. 이를 View로 옮기려면 getter가 꽤나 생기게 되는데 이는 DTO를 적용하면 해결될 수 있으나 DTO로의 변환 코드로 인해 복잡해지는 점과 웹과 달리 View의 경계선이 흐릿한 지금 사용하기에는 잘못 사용할 가능성이 높은 점때문에 추천드리지 않습니다.(특히 시간을 쏟은 만큼의 학습 효과가 있지도 않습니다)
그래서 마음에는 안들지만 이부분이 개선하신다면 getter를 일부 추가하고, View에서만 제한적으로 사용하는 걸 추천드립니다.

wannte replied:
이 부분에 대해서 페어와 많이 얘기했던 부분입니다.
View에서 도메인 객체를 전달받고 이를 작업하는 것이 오른지, 아니면, 도메인 로직에서 이를 다 처리해서 보여줄지에 관해 고민했습니다.
저희 나름의 결론의 근거는, view와 model는 서로 분리되어 있다는 개념을, 서로가 어떻게 구성되어 있는지를 서로 모르는, 즉, view에서는 단순히 출력을 위한 값만을 전달받고 이를 보여준다고 생각을 하였습니다.
휴의 설명대로 라면, 뷰에 경우에는 model을 객체로 전달받아 사용하는 것이 더 바람직한 방법으로 생각이 되네요! 제가 이해한게 맞는지 알고싶습니다 ㅎ

Hue9010 replied:
옆에서 바로 피드백을 주면서 할 수 있다면 DTO를 만들어보라고 하고 싶긴한데... 혼란이 가중 될 것 같군요. 🤔 
도메인 로직에서 이를 다 처리해서 보여주면 View에서 표현 형태가 바뀌면 도메인의 변경이 필요해집니다.
[]로 묶던걸 ()로 바뀐다고 생각해보죠. View에서 어떻게 보여줄 지에 대한 변경인데 LottoStatistics의 변경이 필요해지죠.
하지만 View에서 어떻게 보여줄지(표현할지) 처리를 한다면 도메인의 변경은 필요치 않습니다.

그럼 이게 model 객체를 전달받아 사용한다고 생각하면 될지에 대해서는 No입니다. 웹 처럼 View와 Controller의 경계가 확실히 구분되어 있으면 어렵지 않은데 지금 처럼 같은 자바 환경이라면 구 경계를 느끼기 힘들 수 있는데 데이터가 전달되었다고 생각하는게 좋습니다. 그래서 View에서 모델은 참조하지만 단순 getter로 데이터를 가져와 사용하는 형태로 사용하는 걸 추천드립니다.
즉, View에서는 모델을 DTO라고 생각하고 사용하는걸 추천드려요.

혹시 웹 프로그래밍 경험이 있으시다면 경험해보셨을 수도 있을텐데 어떤 언어로 이루어진 서버던 화면(클라이언트, 웹페이지) 단에서는 서버로부터 받은 `데이터`를 (주로)js를 가지고 처리를 하죠. 서버와 클라이언트 간에는 데이터를 주고 받지 도메인 모델을 주고받지 않습니다. 실제로 사용하지도 못하고요.

다만, 아직은 View까지 자바 환경이고 그 경계를 느끼기 어렵기 때문에 어느정도는 너프하게 View를 구현하는 것도 좋습니다. 웹으로 넘어가면 그 경계가 워낙 명확해지기 때문에 고민 할 것도 적으실겁니다. 😉

Hue9010 commented on src/main/java/lotto/domain/lotto/Lotto.java:
```diff
@@ -0,0 +1,65 @@
+    JDK HashSet 의 생성자에 경우 dummy 를 추가하는 방식으로 생성자 오버로딩을 구현해서 이런식으로 구현해봤습니다.
+    현업에서는 어떻게 사용되는지 궁금합니다!
+     */
+    public Lotto(final List<LottoNumber> numbers, final boolean dummy) {
```
외부에서 사용하지 않네요. 그럼 private로 😉

Hue9010 replied:
> List List를 동시에 사용했을 때, Erase~~ 에러를 띄움. 기존에는 static createByName(List numbers)와 같은 방식으로 구현을 진행했는데, JDK HashSet 의 생성자에 경우 dummy 를 추가하는 방식으로 생성자 오버로딩을 구현해서 이런식으로 구현을 진행.
>
> 현업에서는 어떻게 사용되는지 궁금합니다!

exception을 생각하고 있었는데 컴파일 레벨에서의 에러였군요. 😄
오버로딩의 문제입니다. 같은 시그니처를 가진 생성자나 메서드를 둘 순 없습니다. dummy가 존재하는 생성자는 다른 곳에서 사용하지 않고 있는데 생성자를 두개로 분리할 필요가 있을까요? 이건 분리해야하는 이유를 들어봐야 좀더 얘기를 할 수 있을 것 같아요.

HashSet 내부가 저렇게 구현되었는지는 몰랐군요. 😮 덕분에 새로운걸 알게됐네요. 👍

현업이라기보다는 지금처럼 동일한 파라미터를 가지는 생성자를 만들지 않을 것 같은데 혹여 HashSet처럼 특정 타입으로 만들 필요가 있다면 별도 팩터리를 만들거나 정적 팩터리 메서드를 제공할 것 같습니다.

wannte replied:
한 가지 궁금한 점이 있습니다!
Lotto객체의 경우 List<LottoNumber>의 값을 포함하는 객체인데요!
그렇다면 기본 생성자는 Lotto(List<LottoNumber>)로 정의하는 것이, 정해진 규칙인가요?
아니면 이와 상관 없이, List<Integer>를 받고 이에 따른 List<LottoNumber>를 생성하여 정의하는 것만을 정의하는 방법도 큰 문제가 없을까요?(List<LottoNumber> 말고 오직 List<Integer>만을 이용하여)

개인적인 생각으로는, List<LottoNumber>로 정의하고 다른 방식의 생성자는 이 생성자를 this()로 호출하는 방식으로 사용해야 한다는 생각을 해요. List<Integer>의 경우에는 Eraser로 인한 compile Error 때문에, 정적 팩토리 메소드를 사용하는 것이 바람직한 것 같구요!

휴의 의견 듣고싶어요 ㅎㅎ! 감사합니다.

Hue9010 replied:
우선 컴파일 에러라고 생각하시기보다는 동일한 시그니처를 가진 생성자를 못가진다고 생각해주시는게 좋을 것 같습니다.

기본 생성자는 아무 파라미터도 없는
```java
public Lotto();
```
를 의미합니다. 하나의 생성자만 있다고 그거 기본 생성자라고 칭하진 않습니다!
그렇기 때문에 `기본 생성자는 Lotto(List)로 정의하는 것이, 정해진 규칙인가요?`란 질문은 성립하지 않아요!
약간 말을 바꿔 써보자면 가장 기본적인 생성자를 `Lotto(List)`로 만들 수는 있을 것 같습니다. 가장 기본적이라고 표현한 이유는(우선 이런 표현은 없습니다) 다른 모든 생성자가 내부적으로 this를 통해 이 생성자를 호출하게 사용할 수 있을 것 같네요.

그 이유는 모든 생성자가 각자 생성하는 코드를 가지고 있으면 validation 등 공통적으로 챙겨야 할 코드도 중복적으로 들어가게 되고 실수로 특정 생성자는 공통되어야 할 코드가 빠질 수도 있습니다. 그렇기 때문에 기본적으로 사용할 하나의 생성자를 두고 다른 생성자들은 this를 통해 해당 생성자를 호출해서 사용하는게 좋을 것 같아요.

Hue9010 commented on src/main/java/lotto/domain/lottomachine/LottoMachine.java:
```diff
@@ -0,0 +1,9 @@
+
+import java.util.List;
+
+@FunctionalInterface
```
`@FunctionalInterface`까지 완벽하네요 👍

Hue9010 commented on src/test/java/lotto/domain/LottoManagerTest.java:
```diff
@@ -0,0 +1,63 @@
+    @DisplayName("로또 구매 결과 확인")
+    void buyLottoCheck() {
+        int count = 14;
+        lottoRepository.generateLottoByTicket(() -> Arrays.asList(1, 2, 3, 4, 5, 6), count);
```
테스트에서 많이 사용된다면 매번 람다식을 사용하지 말고 test package 내에 1~6의 숫자만 반환하는 LottoMachine을 만들어 사용할 수도 있습니다. 😉

Hue9010 commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,62 @@
+        return new Ticket(new Money(money));
+    }
+
+    public WinningLotto buyWinningLotto() {
```
전체적으로 메서드의 접근 제어자들을 정리해보죠. 😉

Hue9010 commented on src/main/java/lotto/domain/lotto/Lotto.java:
```diff
@@ -0,0 +1,65 @@
+        Set<LottoNumber> set = new HashSet<>(numbers);
+        set.addAll(lotto.numbers);
+        return totalCount - set.size();
+    }
```
set을 활용한 비교 👍
이게 가능한 이유는 LottoNumber에서 equals, hashCode를 재정의해서죠. 😉 
[Java equals()와 hashCode()에 대해](https://nesoy.github.io/articles/2018-06/Java-equals-hashcode)
[이펙티브 자바 3판 - 아이템 11. equals를 재정의하려거든 hashCode도 재정의하라](https://madplay.github.io/post/always-override-hashcode-when-you-override-equals)

Hue9010 commented on src/main/java/lotto/domain/lotto/LottoRepository.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.Collections;
+import java.util.List;
+
+public class LottoRepository {
```
Repository를 두신 이유가 무엇일까요? 🤔
저장을 하고 있긴 하지만 List로 저장하기 때문에 여러가지 문제가 존재합니다.
크게 두가지 문제가 보이네요. 
- 여러명의 사용자가 이용한다면 각 사용자 별로 자신이 구매한 로또에 대한 구분이 가능할까요?
- 동시에 여러명이 플레이해도 문제가 없을까요? 
  - 한번에 한명씩만 이용한다해도 애플리케이션이 한번 동작 후 종료되지 않고 두번, 세번 돌아도 문제 없을까요??

Controller에서의 지역변수로도 처리가 가능했을 것 같은데 Repository의 존재가 필요한지 생각해볼 수 있을 것 같아요.
(위에 언급한 문제가 지역변수로 처리했을 때도 일부는 존재하지만 요구사항의 스펙상 n회차 진행하는 건 없으나 DB를 대체할 스펙을 추가하셔서 이에 대한 피드백을 남깁니다!)

wannte replied:
구성할 때, Repository로 묶는다는 생각은, 일급 컬렉션으로 관리하고자 했던 것 같아요!
지난 과제에서의 Cars를 묶은 것과 같은 역할을 해줄 것이라고 생각했습니다.
(사실, 묶은 구체적인 근거는 없는 것 같아요.. 단지 List로 관리되는 것은 전부다 별개의 클래스를 만들어서 이를 관리하고자 했던 것 같아요!)

지금 와서 보니 Cars와는 분명한 차이점이 있네요! Cars에 경우에는 묶은 것들 사이에서의 연관관계를 정의해야하지만(maxPosition 등), 이 경우에는 단순히 lotto들의 집합 그 이상 그 이하도 아니네요!

조금 더 근거 있는 리팩토링 진행 해야 될 것이라고 생각이 듭니다!

이렇게 관리했을 때 발생할 수 있는 문제점들을 고민해 보는 것도 좋은 방식 같네요!

Hue9010 replied:
일급 컬렉션과 Repository는 서로 달라요!
Repository는 DB라고 생각해주시는게 좋을 것 같아요. 사전 과제때 컬렉션만 가진 XXXRepository 관련 코드가 있었던걸로 알고 있는데 사전과제 단계에서 실제 DB를 연결하고 사용하게 한다면 많은 에로사항이 있기때문에... 이를 대체하는 형태로 제공하지 않았을까 싶어요.

그렇기 때문에 일급컬렉션이 필요한가 아닌가는 Repository와 별개로 생각해주시는게 좋습니다.

Hue9010 commented on src/main/java/lotto/domain/statistics/LottoStatistics.java:
```diff
@@ -0,0 +1,56 @@
+            }
+            sum += rating.getReward() * ratingCounter.get(rating);
+        }
+        return sum;
```
stream을 활용해 볼 수 있을 것 같네요. (filter 😉 )

Hue9010 replied:
혹은 Rating에서 당첨에 해당하는 Rating만 포함된 List를 반환 할수도 있겠네요.

Hue9010 commented on src/main/java/lotto/domain/primitive/Money.java:
```diff
@@ -0,0 +1,22 @@
+package lotto.domain.primitive;
+
+public class Money {
```
옛날 제가 로또 할 땐 힘들게 느꼈는데...🙄 원시값을 감싸주는건 다들 잘 해주시네요. 👍 💯

Hue9010 commented on src/main/java/lotto/domain/rating/Counter.java:
```diff
@@ -0,0 +1,41 @@
+
+    private final int count;
+
+    public Counter() {
+        this(0);
+    }
```
외부에서 카운터 생성시 0을 받아오는게 아니라 무조건 0 값으로만 생성하게 해두셨네요. 👍👍👍👍
그렇다면 public Counter(final int count)는  외부서 사용하지 못하게 하면 더 완벽해 질 것 같아요.

wannte replied:
테스트에서의 사용을 위해 public으로 설정하는 것은, 부적절한 사용이겠지요?

Hue9010 replied:
넵! 테스트를 위한 코드 추가는 지양합니다. 그렇긴해도 학습 단계에서 어느정도까지 규칙을 지킬지에 대해서는 약간의 자율이라고 생각됩니다.
우선 여긴 getter를 통해서 확인이 가능해 보였습니다. 😉

Hue9010 commented on src/main/java/lotto/domain/lottomachine/RandomLottoMachine.java:
```diff
@@ -0,0 +1,26 @@
+    private final List<Integer> lottoNumbers = IntStream
+            .rangeClosed(LOTTO_MIN_NUMBER, LOTTO_MAX_NUMBER)
+            .boxed()
+            .collect(Collectors.toList());
```
기본 생성자를 만들어 초기화 해줄 수 있지 않을까요? 🤔

Hue9010 commented on src/main/java/lotto/domain/lottomachine/RandomLottoMachine.java:
```diff
@@ -0,0 +1,26 @@
+    public List<Integer> generate() {
+        List<Integer> numbers = new ArrayList<>(lottoNumbers);
+        Collections.shuffle(numbers);
+        return new ArrayList<>(numbers.subList(0, LOTTO_NUMBER_COUNT));
```
굉장히 심플하게 구현해주셨네요. 👍 💯

Hue9010 commented on src/main/java/lotto/domain/rating/LottoResult.java:
```diff
@@ -0,0 +1,38 @@
+    public LottoResult(int matchedCount, boolean hasBonusBall) {
+        this.matchedCount = matchedCount;
+        boolean isSecond = false;
+        if (matchedCount == 5) {
```
5는 매직넘버로 보이네요. 😉

Hue9010 commented on src/main/java/lotto/domain/rating/Rating.java:
```diff
@@ -0,0 +1,37 @@
+package lotto.domain.rating;
+
+public enum Rating {
+    FIRST(2_000_000_000, new LottoResult(6, false)),
```
`_` 👍

Hue9010 commented on src/main/java/lotto/Application.java:
```diff
@@ -0,0 +1,12 @@
+package lotto;
```
> ## Domain 배치에 관한 고민 - 2
> 처음 구현을 진행했을 때, domain 내부적으로는 패키지 구분없이 클래스를 계속 생성하여 프로젝트를 진행.
> 
> 이를 분리를 시킬때, 그 기준을 정하는데에 어려움을 겪음.
> 
> 현재 구성상, 기능적으로 묶이는 것들로 묶고, 특정기능에 속하기 애매한 Primitive타입을 구분.
> 
> 이와 관련한 가이드라인이나 레퍼런스를 찾아 공부를 진행해야겠음!

저도 생각해보니 패키지에 대한 가이드라인을 본적이 드물군요. 🤔
이럴 때는 유명 오픈소스를 참조하는 것도 방법입니다.

lotto.domain 경로에 있기 때문에 lotto.domain.lotto로 두기보다는 이미 lotto 관련 도메인이라는 전제로  domain path에 두는 건 어떨까 싶습니다.

Hue9010 commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    public static void printLottoResults(LottoRepository lottoRepository) {
+        StringBuilder log = new StringBuilder();
```
StringBuilder를 사용하실 때 log라는 변수명으로 사용하고 계신데 어떤 의도인지 궁금합니다!
로그 성격의 문자열이 아닌것 같아서요. 🤔

wannte replied:
log -> buffer는 어떨까요?

Hue9010 commented on src/test/java/lotto/domain/lotto/LottoTest.java:
```diff
@@ -0,0 +1,60 @@
+    @Test
+    @DisplayName("잘못된 개수의 로또 번호")
+    void generateIllegalNumberCountLotto() {
+        assertThatThrownBy(() -> new Lotto(Arrays.asList(1, 2, 3, 4)))
```
실패하는 테스트 👍👍

Hue9010 commented on src/test/java/lotto/domain/primitive/LottoNumberTest.java:
```diff
@@ -0,0 +1,19 @@
+public class LottoNumberTest {
+
+    @ParameterizedTest
+    @ValueSource(ints = {0, -1, 46, 99})
```
경계값 체크 👍

Hue9010 commented on src/test/java/lotto/domain/statistics/LottoStatisticsTest.java:
```diff
@@ -0,0 +1,53 @@
+class LottoStatisticsTest {
+    private static final String ENTER = System.lineSeparator();
+    RatingCounter ratingCounter;
+    LottoStatistics lottoStatistics;
```
접근 제어자가 빠졌네요. 😉

Hue9010 reviewed (changes requested):
안녕하세요 완태! 이번 리뷰를 맡은 휴입니다! 😄 
전체적으로 잘 만들어주셨는데 몇가지 개선하면 좋을 것 같은 내용 피드백 남겼어요.
한번 확인하시고 반영하면 좋을 것 같아요.
궁금한건 언제들 코멘트나 DM으로 문의 주세요. 😉

wannte committed:
style: 상수 inline 처리. ENTER->NEW_LINE

wannte committed:
refactor(LottoStatistics): OutputView 에 객체를 전달.

printLottoResults 에 경우, 모든 값을 버퍼로 입력받고 마지막에만, System.out.print()
printWinningDetail 에 경우에는 각 라인별로 System.out.print()

wannte committed:
refactor(Lotto): 로또 생성자 by int -> static메소드로

wannte committed:
refactor(TestLottoMachine): 테스트용 별도 분리하여 정의

wannte committed:
fix(LottoController): 메서드 접근자 정리

wannte committed:
refactor(LottoStatistics): if문 삭제

MISS일 때, reward가 0기 때문에 따로 예외처리를 할 필요가 없음.

wannte committed:
refactor(LottoResult): 매직 넘버 분리

분리는 했지만, 이 구성이 확정성에 너무 취약한 구성이라, 이를 어떻게 개선할 수 있을까에 대한 고민이 필요

wannte committed:
refactor(Counter): 생성자 보호

wannte committed:
refactor(RandomLottoMachine): 생성자 분리

wannte committed:
refactor: 접근 제어자 추가

wannte committed:
refactor(LottoController): LottoRepository, LottoManager 삭제 및 이동

기존의 테스트들 이전 해야함. 하지만, LottoController를 테스트 하기 힘든 코드가 되어버림

wannte committed:
refactor(LottoController): LottoRepository, LottoManager 삭제 및 이동

기존의 테스트들 이전 해야함. 하지만, LottoController를 테스트 하기 힘든 코드가 되어버림

wannte committed:
refactor(LottoMachine): List<LottoNumber> 생성으로 변경

로또 머신이기때문에, List<LottoNumber>를 하는 것이 더 자연스러운 구현이라고 생각.

wannte committed:
refactor(LottoService): lottoController에서 로직 분리. 통계로직 흡수

wannte committed:
refactor: 전체 구조 정리

wannte committed:
refactor: View -> non static

wannte committed:
style(outputView): getMessage -> printMessage

wannte committed:
refactor: 전체 구조 수정

Hue9010 commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -0,0 +1,74 @@
+            stringBuilder.append(", 보너스 볼 일치");
+        }
+        stringBuilder.append(String.format(" (%d원) - %d개", rating.getReward(), count));
+        System.out.println(stringBuilder.toString());
```
System.out.printX에서는 toString을 안써도 됩니다!

Hue9010 commented on src/main/java/lotto/domain/rating/Counter.java:
```diff
@@ -0,0 +1,41 @@
+        this(0);
+    }
+
+    private Counter(final int count) {
```
👍

Hue9010 commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,66 @@
+import lotto.view.OutputView;
+
+public class LottoController {
+    InputView inputView;
+    OutputView outputView;
```
private를 기본적으로 붙이고 필요할 때 public으로 바꾼다고 생각하면 편합니다. 😉

Hue9010 commented on src/main/java/lotto/domain/LottoService.java:
```diff
@@ -0,0 +1,52 @@
+        return totalSum() / money.getValue();
+    }
+
+    public double totalSum() {
+        double sum = 0;
+        for (Rating rating : Rating.values()) {
+            sum += rating.getReward() * ratingCounter.get(rating);
+        }
+        return sum;
+    }
```
ratingCounter의 총합인데 RatingCounter의 책임이 아닐까요?

Hue9010 commented on src/main/java/lotto/domain/LottoService.java:
```diff
@@ -0,0 +1,52 @@
+
+    public void scratchLotto(WinningLotto winningLotto) {
+        for (Lotto lotto : lottos) {
+            int match = winningLotto.compareLottoNumber(lotto);
+            boolean hasBonusBall = winningLotto.compareBonusBall(lotto);
+            ratingCounter.update(Rating.getRating(match, hasBonusBall));
```
winningLotto.scratch(lotto)로 결과(Rating)를 구하는 책임을 WinningLotto한테 부여할 수 있지 않을까요?

wannte replied:
이 부분은 정말 좋은 책임 이전 같습니다 ㅎㅎ 감사합니다!

Hue9010 reviewed (approved):
완태! 반영을 잘해주셨네요. 👍
이번 단계는 여기서 머지할게요. 간단한 피드백 남겼는데 다음 단계때 반영하면 좋을 것 같습니다.
궁금한건 언제든 질문 주세요!(평일엔 저녁시가에 답변 가능 할 것같아요. 😢 )
주말동안 고생하셨어요! 😉
